docs/src/main/asciidoc/config-mappings.adoc

```diff
@@ -0,0 +1,458 @@
+`GreetingExtraConfiguration` then the resulting default prefix would be `greeting-extra`.
+<2> `!` will be the default value if `greeting.suffix` is not set.
+
+Inject the `GreetingResource` with CDI `@Inject`:
+
+[source,java]
+----
+@Inject
+GreetingConfiguration greetingConfiguration;
+----
+
+Another alternative style provided by Quarkus is to create `GreetingConfiguration` as an interface:
+
+[source,java]
+----
+package org.acme.config;
+
+import io.quarkus.arc.config.ConfigProperties;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import java.util.Optional;
+
+@ConfigProperties(prefix = "greeting")
+public interface GreetingConfiguration {
+
+    @ConfigProperty(name = "message") <1>
+    String message();
+
+    @ConfigProperty(defaultValue = "!")
+    String getSuffix(); <2>
+
+    Optional<String> getName(); <3>
+}
+----
+<1> The `@ConfigProperty` annotation is needed because the name of the configuration property that the method
+corresponds does not follow the getter method naming conventions.
+<2> In this case since `name` is not set, the corresponding property will be `greeting.suffix`.
+<3> It is unnecessary to specify the `@ConfigProperty` annotation because the method name follows the getter method
+naming conventions (`greeting.name` being the corresponding property) and no default value is required.
+
+When using `@ConfigProperties` on a class or an interface, if the value of one of its fields is not provided, the
+application startup will fail, and a `javax.enterprise.inject.spi.DeploymentException` will be thrown. This does not
+apply to `Optional` fields and fields with a default value.
+
+=== Additional notes on @ConfigProperties
+
+When using a regular class annotated with `@ConfigProperties` the class doesn't necessarily have to declare getters and
+setters.Having simple public non-final fields is valid as well.
+
+Furthermore, the configuration classes support nested object configuration. Suppose there was a need to have an extra
+layer of greeting configuration named `content` that would contain a few fields:
+
+[source,java]
+----
+@ConfigProperties(prefix = "greeting")
+public class GreetingConfiguration {
+
+    public String message;
+    public String suffix = "!";
+    public Optional<String> name;
+    public ContentConfig content; <1>
+
+    public static class ContentConfig {
+        public Integer prizeAmount;
+        public List<String> recipients;
+    }
+}
+----
+<1> The name of the field (not the class name) will determine the name of the properties that are bound to the object.
+
+Setting the properties would occur in the normal manner:
+
+.application.properties
+[source,properties]
+----
+greeting.message = hello
+greeting.name = quarkus
+greeting.content.prize-amount=10
+greeting.content.recipients=Jane,John
+----
+
+Furthermore, classes annotated with `@ConfigProperties` can be annotated with Bean Validation annotations:
```

Comment:
I'm a bit surprised that you describe far more features for the deprecated way of doing things?

Reply:
Sure. Let me add some more detail.